src/engine/BoundsExplainer.cpp

```diff
@@ -0,0 +1,270 @@
+{
+	return _varsNum;
+}
+
+BoundsExplainer& BoundsExplainer::operator=( const BoundsExplainer& other )
+{
+	if ( this == &other )
+		return *this;
+
+	if ( _varsNum != other._varsNum )
+	{
+		_upperBoundsExplanations.resize( other._varsNum );
+		_upperBoundsExplanations.shrink_to_fit();
+
+		_lowerBoundsExplanations.resize( other._varsNum );
+		_lowerBoundsExplanations.shrink_to_fit();
+	}
+
+	_rowsNum = other._rowsNum;
+	_varsNum = other._varsNum;
+
+	for ( unsigned i = 0; i < _varsNum; ++i )
+	{
+		if ( _upperBoundsExplanations[i].size() != other._upperBoundsExplanations[i].size() )
+		{
+			_upperBoundsExplanations[i].clear();
+			_upperBoundsExplanations[i].resize( other._upperBoundsExplanations[i].size() );
+			_upperBoundsExplanations[i].shrink_to_fit();
+		}
+
+		if ( _lowerBoundsExplanations[i].size() != other._lowerBoundsExplanations[i].size() )
+		{
+			_lowerBoundsExplanations[i].clear();
+			_lowerBoundsExplanations[i].resize( other._lowerBoundsExplanations[i].size() );
+			_lowerBoundsExplanations[i].shrink_to_fit();
+		}
+
+		std::copy( other._upperBoundsExplanations[i].begin(), other._upperBoundsExplanations[i].end(), _upperBoundsExplanations[i].begin() );
+		std::copy( other._lowerBoundsExplanations[i].begin(), other._lowerBoundsExplanations[i].end(), _lowerBoundsExplanations[i].begin() );
+	}
+
+	return *this;
+}
+
+const std::vector<double>& BoundsExplainer::getExplanation( const unsigned var, const bool isUpper )
+{
+	ASSERT ( var < _varsNum );
+	return isUpper ? _upperBoundsExplanations[var] : _lowerBoundsExplanations[var];
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper )
+{
+	if ( !row._size )
+		return;
+	updateBoundExplanation( row, isUpper, row._lhs );
+}
+
+void BoundsExplainer::updateBoundExplanation( const TableauRow& row, const bool isUpper, const unsigned var )
+{
+	if ( !row._size )
+		return;
+
+	ASSERT ( var < _varsNum );
+	double curCoefficient, ci = 0, realCoefficient;
+	bool tempUpper;
+	unsigned curVar;
+	if ( var == row._lhs )
+		ci = -1;
+	else
+	{
+		for ( unsigned i = 0; i < row._size; ++i )
+			if ( row._row[i]._var == var )
+			{
+				ci = row._row[i]._coefficient;
+				break;
+			}
+	}
+
+	ASSERT( !FloatUtils::isZero( ci ) );
+	std::vector<double> rowCoefficients = std::vector<double>( _rowsNum, 0 );
+	std::vector<double> sum = std::vector<double>( _rowsNum, 0 );
+	std::vector<double> tempBound;
+
+	for ( unsigned i = 0; i < row._size; ++i )
+	{
+		curVar = row._row[i]._var;
+		curCoefficient = row._row[i]._coefficient;
+		if ( FloatUtils::isZero( curCoefficient ) || curVar == var ) // If coefficient is zero then nothing to add to the sum, also skip var
+			continue;
+
+		realCoefficient = curCoefficient / -ci;
+		if ( FloatUtils::isZero( realCoefficient ) )
+			continue;
+
+		tempUpper = ( isUpper && realCoefficient > 0 ) || ( !isUpper &&  realCoefficient < 0 ); // If coefficient of lhs and var are different, use same bound
```

Comment:
I didnt understand the comment.
Also, place comments above the line they are referring to, not at the end of that line